drivers/soundwire/intel_init.c

```diff
@@ -244,6 +244,15 @@ static struct sdw_intel_ctx
 
 		link->cdns = dev_get_drvdata(&ldev->auxdev.dev);
 
+		if (!link->cdns) {
```

Comment:
this is odd @bardliao. If the probe fails, wouldn't the auxiliary_device_add() fail in the first place? Or is this again the case where the probe errors are squelched in drivers/base/dd.c?

Reply:
> this is odd @bardliao. If the probe fails, wouldn't the auxiliary_device_add() fail in the first place? Or is this again the case where the probe errors are squelched in drivers/base/dd.c?

@plbossart The return value is reset to 0 in [really_probe()](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/base/dd.c#L494). See
```
	/*
	 * Ignore errors returned by ->probe so that the next driver can try
	 * its luck.
	 */
```
[ret = 0;](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/base/dd.c#L651)

Reply:
thanks @bardliao that's what I thought indeed.

But what this means is that you are using the drvdata as a proof of success, but the intel_link_probe() sets this pointer *before* an error case. It should be done just before returning.


````diff
diff --git a/drivers/soundwire/intel.c b/drivers/soundwire/intel.c
index 22cb5629f15b..035fbf4f1b4a 100644
--- a/drivers/soundwire/intel.c
+++ b/drivers/soundwire/intel.c
@@ -1360,9 +1360,6 @@ static int intel_link_probe(struct auxiliary_device *auxdev, const struct auxili
        sdw_intel_ops.read_prop = intel_prop_read;
        bus->ops = &sdw_intel_ops;
 
-       /* set driver data, accessed by snd_soc_dai_get_drvdata() */
-       dev_set_drvdata(dev, cdns);
-
        /* use generic bandwidth allocation algorithm */
        sdw->cdns.bus.compute_params = sdw_compute_params;
 
@@ -1382,6 +1379,9 @@ static int intel_link_probe(struct auxiliary_device *auxdev, const struct auxili
         */
        bus->prop.err_threshold = 0;
 
+       /* set driver data, accessed by snd_soc_dai_get_drvdata() */
+       dev_set_drvdata(dev, cdns);
+
        return 0;
 }
````

[intel.diff.txt](https://github.com/thesofproject/linux/files/6091576/intel.diff.txt)

Reply:
@plbossart I am not sure if we need this change. [dev_set_drvdata(dev, NULL);](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/base/dd.c#L626) is called when fail to probe. So `if (!link->cdns)` will still be true even if we call `dev_set_drvdata(dev, cdns);` before an error case.

Reply:
good objection @bardliao thanks for the comment.